bypros/storage/sqlstore/mod.go

```diff
@@ -0,0 +1,230 @@
+package sqlstore
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"os"
+	"sync"
+
+	"go.dedis.ch/cothority/v3/bypros/storage"
+	"go.dedis.ch/cothority/v3/skipchain"
+	"golang.org/x/xerrors"
+)
+
+// Registry is used by the test to cleanly stop all db connections.
+var Registry ConnRegistry = ConnRegistry{}
+
+type ConnRegistry struct {
+	sync.Mutex
+	conns []*sql.DB
+}
+
+func (r *ConnRegistry) add(conn *sql.DB) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.conns = append(r.conns, conn)
+}
+
+func (r *ConnRegistry) StopAll() {
+	r.Lock()
+	defer r.Unlock()
+	for _, db := range r.conns {
+		db.Close()
+	}
+}
+
+// NewSQL returns a new SQL storage.
+func NewSQL() (storage.Storage, error) {
+	dbURL := os.Getenv("PROXY_DB_URL")
+	if dbURL == "" {
+		return nil, xerrors.Errorf("failed to get DB URL: did you set PROXY_DB_URL ?")
+	}
+
+	dbURLRO := os.Getenv("PROXY_DB_URL_RO")
+	if dbURLRO == "" {
+		return nil, xerrors.Errorf("failed to get DB URL: did you set PROXY_DB_URL_RO ?")
+	}
+
+	db, err := sql.Open("pgx", dbURL)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to open db: %v", err)
+	}
+
+	Registry.add(db)
+
+	dbRo, err := sql.Open("pgx", dbURLRO)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to open dbRo: %v", err)
+	}
+
+	Registry.add(dbRo)
+
+	return SQL{
+		db:   db,
+		dbRo: dbRo,
+	}, nil
+}
+
+// SQL defines a storage based on an SQL database.
+//
+// - implements storage.Storage
+type SQL struct {
+	db   *sql.DB
+	dbRo *sql.DB
+}
+
+func (s SQL) Close() {
+	s.db.Close()
+	s.dbRo.Close()
+}
+
+// GetBlock implements storage.Storage. It returns -1 if the block is not found,
+// otherwise its primary key.
+func (s SQL) GetBlock(blockHash []byte) (int, error) {
+	var blockID int
+
+	query := `SELECT block_id 
+	FROM cothority.block 
+	WHERE hash=$1`
+
+	err := s.db.QueryRow(query, blockHash).Scan(&blockID)
+
+	if err == nil {
+		return blockID, nil
+	}
+	if err != sql.ErrNoRows {
+		return -1, xerrors.Errorf("failed to fetch block: %v", err)
+	}
+
+	return -1, nil
+}
+
+// StoreBlock implements storage.Storage. It stores the block and returns its
+// primary key.
+func (s SQL) StoreBlock(block *skipchain.SkipBlock) (int, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	sqlTx, err := s.db.BeginTx(ctx, &sql.TxOptions{
+		Isolation: sql.LevelReadCommitted,
+		ReadOnly:  false,
+	})
+	if err != nil {
+		return -1, xerrors.Errorf("failed to begin transaction: %v", err)
+	}
+
+	storeTx := storeTx{
+		sqlTx: sqlTx,
+		block: block,
+	}
+
+	blockID, err := storeTx.store()
```

Comment:
this API seems very odd to me, as opposed to having `store(block, sqlTx)` 
I really don't see how `storeTx` is not just a function call disguised as an object.